_posts/2020-06-01-rethinking-the-frontend.md

```diff
@@ -0,0 +1,223 @@
+ backend -> db: SQL query
+ db -> backend: result rows
+ backend -> frontend: query results as JSON
+ note left of frontend: frontedn renders\nupdated information
+end
+@enduml
+-->
+![typical SPA sequence](/img/2020-rethinking-frontend/spa-sequence.png)
+
+The above image shows a simplified startup sequence for a hypothetical SPA frontend that shows some
+data queried from a database. There's quite a lot that needs to be done before the frontend is
+ready to render the data the user actually wants to see. The subsequent interaction after the initial
+page had been loaded is much less work.
+
+## A new (old) approach
+
+The SPA approach is good for many applications, but it does add a lot of complexity. In the halcyon
+days of the early 2000s the go to way for web development was fully server side rendered, like [JSP](https://en.wikipedia.org/wiki/JavaServer_Pages) or [PHP](https://www.php.net/).
+There was no API, the server simply rendered full pages and used whatever database resources it needed
+to fill in dynamic parts. JavaScript was mostly used for simple enhancements like form validation before
+sending.
+
+There's nothing wrong with making a fully server rendered application today, but many applications
+have needs and user expectations that require more fine grained updates than reloading the page with
+each user interaction or refreshing the page to check for new updates.
+
+We can combine the server side rendering with [WebSockets](https://en.wikipedia.org/wiki/WebSocket)
+to provide granular updates to the browser and receive callbacks. We can have our cake and eat it too!
+
+## Ripley to the rescue
+
+[Ripley](https://github.com/tatut/ripley) is a new [Clojure](https://clojure.org/) library that
+implements a fully server side rendered programming model that can serve full pages but also
+takes care of updating things that change without page reloads. You can create rich webapps
+without the need for a SPA frontend.
+
+Creating user interfaces with ripley should feel similar to browser side frontend
+development: you create functions that take in data and render HTML output.
+
+```clojure
+;; create a shared state counter (shared by all users)
+(def counter (atom 0))
+
+;; function that takes in counter atom and renders the UI
+(defn counter-app [counter]
+  (h/html
+    [:div
+      "Counter value: " [::h/live {:source counter
+                                   :component #(h/out! (str %))}]
+      [:button {:on-click #(swap! counter inc)} "increment"]
+      [:button {:on-click #(swap! counter dec)} "decrement"]]))
+```
+
+The above example looks similar to [Reagent](https://github.com/reagent-project/reagent)
+(a ClojureScript React wrapper), but it is all rendered on the server.
+The button callbacks defined with `:on-click` are actually run on the server.
+The special `[::h/live ...]` element register a source on the server for this
+page. Whenever the atom changes, the component rerendered on the server and the
```

Comment:
"the component _is_ re-rendered", but using the active voice might make it clearer to know who's doing what: "…changes, Ripley re-renders the component and sends the resulting HTML to the client…"